clippy_lints/src/operators/double_const_comparison.rs

```diff
@@ -0,0 +1,204 @@
+#![allow(clippy::match_same_arms)]
+
+use std::cmp::Ordering;
+
+use clippy_utils::consts;
+use clippy_utils::consts::{ConstEvalLateContext, Constant};
+use if_chain::if_chain;
+use rustc_hir::{BinOpKind, Expr, ExprKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::{layout::HasTyCtxt, Ty, TypeckResults};
+use rustc_span::source_map::{Span, Spanned};
+
+use clippy_utils::diagnostics::span_lint_and_note;
+use clippy_utils::source::snippet;
+use clippy_utils::SpanlessEq;
+
+use super::IMPOSSIBLE_DOUBLE_CONST_COMPARISONS;
+use super::INEFFECTIVE_DOUBLE_CONST_COMPARISONS;
+
+// Extract a comparison between a const and non-const
+// Flip yoda conditionals, turnings expressions like `42 < x` into `x > 42`
+fn comparison_to_const<'tcx>(
+    ctx: &mut ConstEvalLateContext<'_, 'tcx>,
+    typeck: &TypeckResults<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+) -> Option<(CmpOp, &'tcx Expr<'tcx>, &'tcx Expr<'tcx>, Constant, Ty<'tcx>)> {
+    if_chain! {
+        if let ExprKind::Binary(operator, left, right) = expr.kind;
+        if let Ok(cmp_op) = CmpOp::try_from(operator.node);
+        then {
+            match (ctx.expr(left), ctx.expr(right)) {
+                (Some(_), Some(_)) => None,
+                (_, Some(con)) => Some((cmp_op, left, right, con, typeck.expr_ty(right))),
+                (Some(con), _) => Some((cmp_op.reverse(), right, left, con, typeck.expr_ty(left))),
+                _ => None,
+            }
+        } else {
+            None
+        }
+    }
+}
+
+pub(super) fn check<'tcx>(
+    cx: &LateContext<'tcx>,
+    and_op: Spanned<BinOpKind>,
+    left_cond: &'tcx Expr<'tcx>,
+    right_cond: &'tcx Expr<'tcx>,
+    span: Span,
+) {
+    if_chain! {
+        // Ensure that the binary operator is &&
+        if and_op.node == BinOpKind::And;
+
+        let typeck_results = cx.typeck_results();
```

Comment:
nit: we should filter by the type first to avoid doing expensive const computations

Reply:
Thanks for the feedback. This was a bit tricky to implement, at least if we want to detect `a > b` code where `a` and `b` are not the same type. I've pushed a commit that adds test cases for this.

Here's the helper function I came up with for implementing your optimization: (Not committed yet):

```rust
// Check that our binary operation is of the form `a CMP b && c CMP d`,
// and that the same *type* occurs at least once on both sides of `&&`
fn is_double_comparison<'tcx>(
    typeck: &TypeckResults<'tcx>,
    left_cond: &'tcx Expr<'tcx>,
    right_cond: &'tcx Expr<'tcx>,
) -> bool {
    if_chain! {
        if let ExprKind::Binary(left_cmp_op, left_left_op, left_right_op) = left_cond.kind;
        if let ExprKind::Binary(right_cmp_op, right_left_op, right_right_op) = right_cond.kind;

        if CmpOp::try_from(left_cmp_op.node).is_ok();
        if CmpOp::try_from(right_cmp_op.node).is_ok();

        then {
            let left_types = [typeck.expr_ty(left_left_op), typeck.expr_ty(left_right_op)];
            let right_types = [typeck.expr_ty(right_left_op), typeck.expr_ty(right_right_op)];

            left_types.iter().any(|typ| right_types.contains(typ))
        }
        else {
            false
        }
    }
}
```
Honestly, I think we need to benchmark this. The new code adds a fair bit of complexity, and it's not obvious to me that it's a performance win at all. Do you know a good way to benchmark a single lint on real-world code?

Reply:
It's kinda annoying, you can just delete all other lint passes and run it and see. The clippy channel in zulip may know.

(The problem is that rustc will run all lints including disabled ones, as one giant interleaved pass.)

I was going to suggest that we just `expr_ty` the two parts and check against concrete numeric types but I do realize that still needs the splitting. Maybe it doesn't matter, and those operators are used  with numeric types most of the time anyway.

I'm ok merging this with the lint name updated and if you'd like you can explore benchmarking and optimizing it later.

Reply:
I think I could have matched both halves against a list of concrete types, but that might also a bit involved, since the lint also works for string slices, tuples, arrays, basically anything in [Constant](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/consts/enum.Constant.html). I'd like for the lint to keep working on those types, since I think it's easy to makes mistakes with them as well, e.g. the `name < "Jennifer" && name > "Shannon"` test case.

I wasn't able to get any benchmarking to work, so I settled on a middle solution and implemented the simplest possible optimization: Check that both sides of the root `&&` operator are themselves binary operations. That should filter out most expressions before the expensive const checking stuff.